### PR TITLE
[UX] Lead generation analysis for contacts chart

### PR DESCRIPTION
**Pull request by andersonjeccel**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🔴 
| New feature/enhancement? (use the a.x branch)      | 🟢
| Deprecations?                          | 🔴
| BC breaks? (use the c.x branch)        | 🔴
| Automated tests included?              | 🔴 
| Related user documentation PR URL      | mautic/user-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation-new#... 
| Issue(s) addressed                     | Fixes #... 

## Description

This PR helps marketers to have Marketing Insights (MI) about lead generation at a glance, depending on the results and selected date range.

I created an analysis model that shows different insights about the data depending on the time period the user is looking at. If they are looking at an entire month, it talks about things like stability, which were the best days, etc. If they are looking at a week, the system understands that you can't infer everything with so little data. So, it provides other metrics that can be reliable only with what is known from the period. 

Since it's in a marketing system where everything you know is what your company knows, the level of personalization is absurd.
Market trends using industry data may not be as accurate when it comes to the wake-up call like "come on, champ, your company is taking 3 weeks to recover from a strategy failure."
The final idea is for the employee to be able to pass the wake-up call forward.
Until everyone gets a wake-up call and does something about it.

The main goal is to empower marketers with info such as:

- How much time it takes for our company to recover after a big drop in lead generation in long term or how many days within the same week?
- Do we have a concentration of new contacts in certain days while others none?
- What's the average amount of new leads everyday?
- Do we have best days of the week for lead generation?
- Which was the best day this month?
- How much stable is our process?

image

image

image

---
### 📋 Steps to test this PR:

1. Pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester)) using your existing mock data
2. Change the dates on the dashboard to view periods of 60, 30 and 7 days.
3. Insights will disappear when the range is bigger than a month.
4. Monthly insights will be available when viewing ~30 days. Click on MI to see
5. Weekly insights will be available when viewing ~7 days. Click on MI to see
6. To check comparisons: for example, if you imported leads on day 1, set your initial date to day 2 so comparison about previous period becomes available.

---

# Technical details you can skip

## Metrics tree
**Always visible**
Growth rate (comparison with previous period)
Total leads generated

**Monthly**
Average daily leads
Consistency (standard deviation)
Best day of the month and maximum value
High performance days
Growth/drop alert
Lead generation volatility

**Weekly**
Most productive day of the week
Worst days
Alert if the worst weekday from last week is the same this week
General weekly growth direction

**Long term metrics** (well, the best ones but dropped out for now to be included in separate PR)
Weekend trend vs. day of the week
Best day of the week over months
Time to recover lead generation after a big drop
Lead concentration index (if we're depending of leads generated only in part of the week over months)

## MI general structure
- Kind of information title (h2)
- Description
- Content
- Footer info (related to the source of data)

### Total leads generated
A sum of all points on the chart.

Works this way:

1. Checks if the 'showTotal' variable is defined.
2. If 'showTotal' is defined, it calculates the sum of all values in the 'dataValues' array.
3. Displays the total sum in a formatted div element.
4. Also if 'showTotal' is defined, it reduces the chart height by 60 pixels.

So:

- The total count is only shown if 'showTotal' is defined in PHP.
- Uses the 'reduce' function to efficiently sum all values in the 'dataValues' array.
- Modifies the 'chartHeight' to accommodate the added total count display without overlapping other elements
- The calculation is done server-side.

### Growth rate
Value in percentage comparing to the previous period (same amount of days you have in the current range, but in the past).

How it works:

1. Checks if previous period data is available and properly structured.
2. Calculates total values for both current and previous periods using the reduce function.
3. Determines the type of change (increase, decrease, or no change) between the two periods.
4. Calculates the percentage change.
5. Displays the comparison results.

Aspects:

- Only shows comparison if previous period data is available.
- Uses reduce function to sum data values for both periods.
- Computes percentage change, rounded to two decimal places.
- Employs up/down arrows and colors (green for increase, red for decrease) for quick visual feedback.
- Manages scenarios like when previous total is zero or when there's no change (important).

### Consistency (standard deviation)
Measure how consistent your results are. A low standard deviation indicates more predictable results, while a high one suggests greater volatility.
For example, a ±5 value means that, if you generate an average of 25 leads, some days you in fact generate 20 and some days 30.

Then

1. Since it's not that obvious, includes a tooltip with additional information about standard deviation
2. Calculates and displays the standard deviation value, rounded down to the nearest integer

### Most productive day of the week
The chart naturally doesn't show the day of the week, so you have to check the calendar every time an analysis regarding the day of the week become relevant.
This metric helps telling directly which day had the most generated leads amount.

Works like:

1. Iterates through the chart data, grouping the lead counts by day of the week.
2. Calculates the average number of leads for each day of the week.
3. Identifies the day with the highest average.
4. Displays the best day and its average lead count.

It's smart! (relatively):

- It handles the case where there might be no clear best day (e.g., if all days have 0 leads).
- The average is rounded down to the nearest integer to cut off decimal numbers.
- It doesn't make assumptions about which days of the week are present in the data, only working with the days that appear in the dataset.

### General weekly growth direction

A weekly trend that compares the average of the first three days with the average of the last three days of the week, and from this indicates the general direction of the number of leads throughout the week, whether it's growth, decline, or stability.

_and more..._

### CODE REVIEW

**Comment by andersonjeccel** [comment]
Points for improvement:
- [x] show high performance days only when we have more than 1
- [x] hide message when there’s 0% change in lead generation
- [x] Use a different footer message for Weekly insights 
- [x] Consider adding a tooltip for each metric to explain how they are built (or move to documentation)
- [x] improve the clarity of “average” in weekly insights
- [x] how much relatable is to set a fixed range of 6 months for weekly analysis when periods are often affected by special events/holidays like Father’s Day?
- [x] Consider moving weekly metrics to annual 
- [x] Reduce the “lead generation” words repetition

**Review by shinde-rahul — APPROVED** [review]
UX++, thanks @andersonjeccel.

> Code changes LGTM!!!

I'll redo the review once PR is in active state.

**Comment by andersonjeccel** [comment]
I need help to fix the test :/
@abhisekmazumdar 
@LordRembo

File: app/bundles/LeadBundle/Tests/EventListener/DashboardSubscriberTest.php [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+        $this->router        = $this->createMock(RouterInterface::class);
+        $this->translator    = $this->createMock(TranslatorInterface::class);
+        $this->dateHelper    = new DateHelperStub();
+
+        $this->dashboardSubscriber = new DashboardSubscriber(
+            $this->leadModel,
+            $this->leadListModel,
+            $this->router,
+            $this->translator,
+            $this->dateHelper
```
Comment (abhisekmazumdar) [inline app/bundles/LeadBundle/Tests/EventListener/DashboardSubscriberTest.php:81]: The changes to $this->dateHelper will resolve this issue.

**Comment by abhisekmazumdar** [comment]
I feel that after making these changes, most of the test cases will also pass. If not, I will be around.

File: app/bundles/LeadBundle/EventListener/DashboardSubscriber.php [hunk 2]
```diff
@@ -81,21 +81,193 @@ public function onWidgetDetailGenerate(WidgetDetailEvent $event): void
             }
 
             if (!$event->isCached()) {
+                $chartData = $this->leadModel->getLeadsLineChartData(
```
Comment (shinde-rahul) [inline app/bundles/LeadBundle/EventListener/DashboardSubscriber.php:84]: I suggest breaking the long method into smaller methods; this will improve the DX and help with maintainability.

File: app/bundles/LeadBundle/Resources/views/Widget/created_leads_in_time.html.twig
Comment (shinde-rahul) [inline app/bundles/LeadBundle/Resources/views/Widget/created_leads_in_time.html.twig:1]: This needs some indentation.

File: app/bundles/LeadBundle/Tests/EventListener/DashboardSubscriberTest.php [hunk 3]
```diff
@@ -0,0 +1,409 @@
+<?php
+
```
Comment (shinde-rahul) [inline app/bundles/LeadBundle/Tests/EventListener/DashboardSubscriberTest.php:2]: `declare(strict_types=1);` is missing.

File: app/bundles/LeadBundle/Tests/EventListener/DashboardSubscriberTest.php [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+use Mautic\LeadBundle\EventListener\DashboardSubscriber;
+use Mautic\LeadBundle\Model\LeadModel;
+use Mautic\LeadBundle\Model\ListModel;
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+class DateHelperStub
+{
+    public function formatRange(\DateInterval $interval)
```
Comment (shinde-rahul) [inline app/bundles/LeadBundle/Tests/EventListener/DashboardSubscriberTest.php]: A return type is missing. Please check the PHPSTAN pipeline for more details.

File: app/bundles/LeadBundle/Tests/EventListener/DashboardSubscriberTest.php [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+
+use Mautic\DashboardBundle\Event\WidgetDetailEvent;
+use Mautic\LeadBundle\EventListener\DashboardSubscriber;
+use Mautic\LeadBundle\Model\LeadModel;
+use Mautic\LeadBundle\Model\ListModel;
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+class DateHelperStub
```
Comment (shinde-rahul) [inline app/bundles/LeadBundle/Tests/EventListener/DashboardSubscriberTest.php]: We should have one class per file. Please move this to a separate file.

**Review by shinde-rahul — CHANGES_REQUESTED** [review]
I added some suggestions. Please check the failing pipelines logs as well.

File: app/bundles/LeadBundle/Tests/EventListener/DashboardSubscriberTest.php (hunk 5 above)
Comment (abhisekmazumdar) [reply app/bundles/LeadBundle/Tests/EventListener/DashboardSubscriberTest.php]: Should `DateHelperStub.php` be created under `Tests/Services` ?

I also see that `app/bundles/LeadBundle/Tests/Services/PeakInteractionTimerTest.php` contains two classes. I'm not familiar with the standards, but is it acceptable to have two classes in that file?

File: app/bundles/LeadBundle/EventListener/DashboardSubscriber.php (hunk 2 above)
Comment (abhisekmazumdar) [reply app/bundles/LeadBundle/EventListener/DashboardSubscriber.php:84]: Do you suggest converting all the changes from line 84 to 268 into smaller helper methods?

**Comment by RCheesley** [comment]
It looks like we've got quite some code review feedback here, so I am going to move this back to the 6.0 Release Candidate milestone to give us more time to get it over the line.

**Comment by RCheesley** [comment]
Let's get this rebased to 7.x and aim for the alpha release, end of April?

**Comment by RCheesley** [comment]
Can we get this rebased and fix the conflicts? Thanks!

**Comment by andersonjeccel** [comment]
@RCheesley Let me try again, I was stuck (help needed)

**Comment by RCheesley** [comment]
Thanks @andersonjeccel I've updated the base branch to 7.x - did you still need some assistance with this one?

**Comment by andersonjeccel** [comment]
@RCheesley I was able to fix the tests, all good now

**Review by kuzmany — APPROVED** [review]
I did test and it works, only I saw not rounded number and too many tuesdays (image bellow).
I would use LLM for kind of this reports, instead of write too much of code.

image

**Comment by andersonjeccel** [comment]
@kuzmany How to reproduce what you're seeing?

Info like this would help:
- Contacts amount
- Screenshot of the chart

@kuzmany @RCheesley Integrating a LLM would be unfeasible for now... Do you think I should keep working on the feature following the current format or abandon the idea?
